File: class/practice_4.1/pizza.cpp
#include "pizza.hpp"
#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <print>

Pizza::Pizza() : total_price(0.0) {}

void Pizza::addIngredient(const Ingredient& ingredient) {
    ingredients.push_back(ingredient);
    total_price += ingredient.price;
}

double Pizza::getTotalPrice() const {
    return total_price;
}

const std::vector<Ingredient>& Pizza::getIngredients() const {
    return ingredients;
}

PizzaBuilder::PizzaBuilder(const std::string& ingredientsFile) : ingredientsFile(ingredientsFile) {
    loadIngredients();
}

Pizza PizzaBuilder::buildPizza() {
    return pizza;
}

std::vector<std::string> PizzaBuilder::getAvailableIngredients() const {
    std::vector<std::string> ingredientNames;
    for (const auto& pair : ingredients) {
        ingredientNames.push_back(pair.first);
    }
    return ingredientNames;
}

bool PizzaBuilder::addIngredientToPizza(const std::string& ingredientName) {
    auto it = ingredients.find(ingredientName);
    if (it != ingredients.end()) {
        pizza.addIngredient(it->second);
        return true;
    } else {
        return false;
    }
}

void PizzaBuilder::displayAvailableIngredients() const {
    std::println("Available ingredients:");
    for (const auto& pair : ingredients) {
        std::println("- {} (${})", pair.second.name, pair.second.price);
    }
}

void PizzaBuilder::loadIngredients() {
    std::ifstream file(ingredientsFile);
    if (file.is_open()) {
        std::string line;
        while (std::getline(file, line)) {
            std::stringstream ss(line);
            std::string name, priceStr;
            std::getline(ss, name, ',');
            std::getline(ss, priceStr, ',');
            double price = std::stod(priceStr);
            ingredients[name] = { name, price };
        }
        file.close();
    } else {
        std::println("Error opening ingredients file: {}", ingredientsFile);
    }
}

File: home/home_4.1/pizza.swift
import Foundation

struct Ingredient {
    let name: String
    let price: Double
}

class Pizza {
    private var ingredients: [Ingredient] = []
    private var totalPrice: Double = 0.0

    func addIngredient(_ ingredient: Ingredient) {
        ingredients.append(ingredient)
        totalPrice += ingredient.price
    }

    func getTotalPrice() -> Double {
        return totalPrice
    }

    func getIngredients() -> [Ingredient] {
        return ingredients
    }
}

class PizzaBuilder {
    private var ingredients: [String: Ingredient] = [:]
    private var ingredientsFile: String

    init(ingredientsFile: String) {
        self.ingredientsFile = ingredientsFile
        loadIngredients()
    }

    private func loadIngredients() {
        let fileURL = URL(fileURLWithPath: ingredientsFile)
        do {
            let data = try String(contentsOf: fileURL, encoding: .utf8)
            let lines = data.split(separator: "\n")
            for line in lines {
                let parts = line.split(separator: ",")
                if parts.count == 2 {
                    let name = String(parts[0])
                    if let price = Double(parts[1]) {
                        ingredients[name] = Ingredient(name: name, price: price)
                    }
                }
            }
        } catch {
            print("Error loading ingredients from file: \(error)")
        }
    }

    func getAvailableIngredients() -> [String] {
        return Array(ingredients.keys)
    }

    func addIngredientToPizza(pizza: Pizza, ingredientName: String) -> Bool {
        if let ingredient = ingredients[ingredientName] {
            pizza.addIngredient(ingredient)
            return true
        }
        return false
    }

    func buildPizza() -> Pizza {
        return Pizza()
    }

    func getClassicPizzaOptions() -> [String: [Ingredient]] {
        return [
            "Margherita": [
                Ingredient(name: "Tomato", price: 1.5),
                Ingredient(name: "Cheese", price: 2.0)
            ],
            "Pepperoni": [
                Ingredient(name: "Tomato", price: 1.5),
                Ingredient(name: "Cheese", price: 2.0),
                Ingredient(name: "Pepperoni", price: 2.5)
            ],
            "Vegetarian": [
                Ingredient(name: "Tomato", price: 1.5),
                Ingredient(name: "Cheese", price: 2.0),
                Ingredient(name: "Mushroom", price: 1.0),
                Ingredient(name: "Bell Pepper", price: 1.0)
            ]
        ]
    }
}
